src/xref2/link.ml

```diff
@@ -400,16 +415,180 @@ and open_ env parent = function
+        | ModuleSubstitution m -> Env.open_module_substitution m env
+        | Type _ -> env
+        | TypeSubstitution t -> Env.open_type_substitution t env
+        | ModuleType mt -> module_type env mt
+        | ModuleTypeSubstitution mts ->
+            let env = Env.open_module_type_substitution mts env in
+            module_type_substitution env mts
+        | Value _ -> env
+        | Comment _ -> env
+        | TypExt _ -> env
+        | Exception _ -> env
+        | Class _ -> env
+        | ClassType _ -> env
+        | Include i -> include_ env i
+        | Open _ -> env)
+      env s
+
+  and module_type_substitution : Env.t -> ModuleTypeSubstitution.t -> Env.t =
+   fun env m -> module_type_expr env m.manifest
+
+  and include_ : Env.t -> Include.t -> Env.t =
+   fun env i ->
+    let open Include in
+    signature_items env i.expansion.content.items
+
+  and module_type : Env.t -> ModuleType.t -> Env.t =
+   fun env m ->
+    match m.expr with None -> env | Some expr -> module_type_expr env expr
+
+  and module_ : Env.t -> Module.t -> Env.t =
+   fun env m ->
+    let open Module in
+    let env = module_decl env m.type_ in
+    match m.type_ with
+    | Alias (`Resolved _, Some exp) -> simple_expansion env exp
+    | Alias _ | ModuleType _ -> env
+
+  and module_decl : Env.t -> Module.decl -> Env.t =
+   fun env decl ->
+    let open Module in
+    match decl with
+    | ModuleType expr -> module_type_expr env expr
+    | Alias (_, None) -> env
+    | Alias (_, Some e) -> simple_expansion env e
+end
 let rec unit env t =
   let open Compilation_unit in
   let content =
-    match t.content with
-    | Module sg ->
-        let sg = signature env (t.id :> Id.Signature.t) sg in
-        Module sg
-    | Pack _ as p -> p
+    if t.Lang.Compilation_unit.linked || t.hidden then t.content
+    else
+      match t.content with
+      | Module sg ->
+          let sg = signature env (t.id :> Id.Signature.t) sg in
+          Module sg
+      | Pack _ as p -> p
+  in
+  let source_info =
+    let env = Build_env.unit env t in
+    let open Source_info in
+    match t.source_info with
+    | Some inf ->
+        let jump_to v f_impl f_doc =
+          let documentation =
+            match v.documentation with Some p -> Some (f_doc p) | None -> None
+          in
+          let implementation =
+            match v.implementation with
+            | Some (Unresolved p) -> (
+                match f_impl p with
+                | Some x -> Some (Resolved x)
+                | None -> v.implementation)
+            | x -> x
+          in
+          { documentation; implementation }
+        in
+        let infos =
+          List.map
+            (fun (i, pos) ->
+              let info =
+                match i with
+                | Value v ->
+                    Value
+                      (jump_to v
+                         (Shape_tools.lookup_value_path env)
+                         (value_path ~report_errors:false env))
```

Comment:
In the end, there are much less failures than before (that's again because we removed the non-persistent identifiers!).

So, I'm in favor of considering such failures as normal failures, and printing them if the options say so.

(With non-persistent occurrences, there were warnings that one could never get rid of, which is why I introduced the non-error reporting)